File: node/src/drivers/vmm/hostcall_global.go
package vmm

import "encoding/json"

// VmCallback handles appengine host-call packets globally for all appengine
// runtimes (wasm/docker/javascript/elpify/elpian) through the same ZeroMQ callback channel.
func (wm *Vmm) VmCallback(dataRaw string) (string, int64) {
	println(dataRaw)
	data := map[string]any{}
	err := json.Unmarshal([]byte(dataRaw), &data)
	if err != nil {
		println(err)
		return err.Error(), 0
	}
	reqIdRaw, err := checkField(data, "requestId", float64(0))
	if err != nil {
		println(err)
		return err.Error(), 0
	}
	reqId := int64(reqIdRaw)
	key, err := checkField(data, "key", "")
	if err != nil {
		println(err)
		return err.Error(), reqId
	}
	input, err := checkField[map[string]any](data, "input", nil)
	if err != nil {
		println(err)
		return err.Error(), reqId
	}

	switch key {
	case "execDocker", "execVm":
		return wm.handleExecDocker(input, reqId)
	case "copyToDocker", "copyToVm":
		return wm.handleCopyToDocker(input, reqId)
	case "checkTokenValidity":
		return wm.handleCheckTokenValidity(input, reqId)
	case "plantTrigger":
		return wm.handlePlantTrigger(input, reqId)
	case "signalPoint":
		return wm.handleSignalPoint(input, reqId)
	case "runVm":
		return wm.handleRunVM(input, reqId)
	case "terminateVm":
		return wm.handleTerminateVM(input, reqId)
	case "sendMessageOnChain":
		return wm.handleSendMessageOnChain(input, reqId)
	case "createCreature":
		return wm.handleCreatureCrud("create", input, reqId)
	case "updateCreature":
		return wm.handleCreatureCrud("update", input, reqId)
	case "deleteCreature":
		return wm.handleCreatureCrud("delete", input, reqId)
	case "getCreature":
		return wm.handleCreatureCrud("get", input, reqId)
	case "listCreatures":
		return wm.handleCreatureCrud("list", input, reqId)
	case "createResourceStore", "createVmOwnedStore":
		return wm.handleResourceStoreCrud("create", input, reqId)
	case "updateResourceStore", "updateVmOwnedStore":
		return wm.handleResourceStoreCrud("update", input, reqId)
	case "deleteResourceStore", "deleteVmOwnedStore":
		return wm.handleResourceStoreCrud("delete", input, reqId)
	case "getResourceStore", "getVmOwnedStore":
		return wm.handleResourceStoreCrud("get", input, reqId)
	case "listResourceStores", "listVmOwnedStores":
		return wm.handleResourceStoreCrud("list", input, reqId)
	case "createResourceEntity":
		return wm.handleResourceEntityCreate(input, reqId)
	case "deleteResourceEntity":
		return wm.handleResourceEntityDelete(input, reqId)
	case "createWorkchain":
		return wm.handleVmChainRequest("createWorkchain", input, reqId)
	case "deleteWorkchain":
		return wm.handleVmChainRequest("deleteWorkchain", input, reqId)
	case "createSubchain":
		return wm.handleVmChainRequest("createSubchain", input, reqId)
	case "deleteSubchain":
		return wm.handleVmChainRequest("deleteSubchain", input, reqId)
	case "execShellAction":
		return wm.handleExecShellAction(input, reqId)
	case "microGenId":
		return wm.handleMicroHostAction("genId", input, reqId)
	case "microGetLink":
		return wm.handleMicroHostAction("getLink", input, reqId)
	case "microPutLink":
		return wm.handleMicroHostAction("putLink", input, reqId)
	case "microDelKey":
		return wm.handleMicroHostAction("delKey", input, reqId)
	case "microGetJson":
		return wm.handleMicroHostAction("getJson", input, reqId)
	case "microPutJson":
		return wm.handleMicroHostAction("putJson", input, reqId)
	case "microGetByPrefix":
		return wm.handleMicroHostAction("getByPrefix", input, reqId)
	case "microHasAccessToPoint":
		return wm.handleMicroHostAction("hasAccessToPoint", input, reqId)
	case "microSignalUser":
		return wm.handleMicroHostAction("signalUser", input, reqId)
	case "microSignalGroup":
		return wm.handleMicroHostAction("signalGroup", input, reqId)
	case "microJoinGroup":
		return wm.handleMicroHostAction("joinGroup", input, reqId)
	case "log":
		_, err := checkField(input, "text", "")
		if err != nil {
			println(err)
			return err.Error(), reqId
		}
	}

	return "{}", reqId
}

File: node/src/drivers/vmm/hostcall_global.rs
//! Translation of `drivers/vmm/hostcall_global.go`.

use serde_json::Value;

use super::vmm::{check_i64, check_str, Vmm};

impl Vmm {
    /// Entry point: parse a host-call message coming from the appengine over
    /// ZMQ and route it to the right handler. Returns `(result_json,
    /// requestId)` so the caller can produce an `apiResponse` envelope.
    pub fn vm_callback(&self, data_raw: &str) -> (String, i64) {
        let data: Value = match serde_json::from_str(data_raw) {
            Ok(v) => v,
            Err(e) => return (format!("{{\"error\":\"{}\"}}", e), 0),
        };
        let req_id = check_i64(&data, "requestId", 0);
        let key = check_str(&data, "key", "");
        let input = data.get("input").cloned().unwrap_or(Value::Null);

        match key.as_str() {
            "execDocker" | "copyToDocker" | "execVm" | "copyToVm" | "runVm" => {
                ("unsupported runtime".into(), req_id)
            }
            "checkTokenValidity" => self.handle_check_token_validity(&input, req_id),
            "plantTrigger" => self.handle_plant_trigger(&input, req_id),
            "signal" => self.handle_signal_store(&input, req_id),
            "terminateVm" => self.handle_terminate_vm(&input, req_id),
            "sendMessageOnChain" => self.handle_send_message_on_chain(&input, req_id),
            "createCreature" => self.handle_creature_crud("create", &input, req_id),
            "updateCreature" => self.handle_creature_crud("update", &input, req_id),
            "deleteCreature" => self.handle_creature_crud("delete", &input, req_id),
            "getCreature" => self.handle_creature_crud("get", &input, req_id),
            "listCreatures" => self.handle_creature_crud("list", &input, req_id),
            "createResourceStore" | "createVmOwnedStore" => {
                self.handle_resource_store_crud("create", &input, req_id)
            }
            "updateResourceStore" | "updateVmOwnedStore" => {
                self.handle_resource_store_crud("update", &input, req_id)
            }
            "deleteResourceStore" | "deleteVmOwnedStore" => {
                self.handle_resource_store_crud("delete", &input, req_id)
            }
            "getResourceStore" | "getVmOwnedStore" => {
                self.handle_resource_store_crud("get", &input, req_id)
            }
            "listResourceStores" | "listVmOwnedStores" => {
                self.handle_resource_store_crud("list", &input, req_id)
            }
            "createStore" => self.handle_store_crud("create", &input, req_id),
            "updateStore" => self.handle_store_crud("update", &input, req_id),
            "deleteStore" => self.handle_store_crud("delete", &input, req_id),
            "getStore" => self.handle_store_crud("get", &input, req_id),
            "listStores" => self.handle_store_crud("list", &input, req_id),
            "createResourceEntity" => self.handle_resource_entity_create(&input, req_id),
            "deleteResourceEntity" => self.handle_resource_entity_delete(&input, req_id),
            "createWorkchain" | "deleteWorkchain" | "createSubchain" | "deleteSubchain" => {
                self.handle_vm_chain_request(&key, &input, req_id)
            }
            "execShellAction" => self.handle_exec_shell_action(&input, req_id),
            "genId" | "getLink" | "delKey" | "createAccess" | "deleteAccess" | "getJson"
            | "putJson" | "getByPrefix" | "hasAccessToStore" | "signalUser" | "signalGroup"
            | "joinGroup" => self.handle_micro_host_action(&key, &input, req_id),
            "log" | "vmLog" | "buildLog" | "output" | "vmOutput" => {
                self.handle_vm_log_event(&input, req_id)
            }
            _ => ("{}".into(), req_id),
        }
    }
}
